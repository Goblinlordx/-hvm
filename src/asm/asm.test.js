// import { tokenize, parser } from "./";

// const basicTokenizerTests = [
//     ["tokenize skips comment", "// Test Comment"],
//     ["tokenize skips basic comment with whitespace", "   // Test Comment  "],
//     ["tokenizes new lines", ["", ""]],
//     [
//         "tokenizes basic comment in with empty lines",
//         ["", "// Test Comment", "// Test line 2", ""]
//     ],
//     ["tokenizes @ symbol", "@"],
//     ["tokenizes @ symbol with name", "@S.FS-C;"],
//     ["tokenizes positive numeric value", "123"],
//     ["tokenizes positive numeric value with whitespace", " 123 "],
//     ["fail parsing number with decimal", "1.23", true],
//     ["tokenize instruction as text", "D=A-1"],
//     ["tokenizes neg", "-"],
//     [
//         "tokenize basic program",
//         [
//             "// Basic program",
//             "// -- Setup",
//             "  @R.-;1+=",
//             "  D=A",
//             "  @0",
//             "  M=D",
//             "  @2",
//             "  D=A",
//             "  @1",
//             "  M=D",
//             "// Almost max program",
//             "(MAX)",
//             "  @R_0",
//             "  D=M",
//             "  @R.1",
//             "  D=D-M",
//             "  D;JGT",
//             "",
//             "  @-123",
//             "  @-test",
//             "  @0",
//             ""
//         ]
//     ]
// ];

// basicTokenizerTests.forEach(([tag, data, expectThrow]) => {
//     it(tag, () => {
//         const wrap = () => {
//             const d = data instanceof Array ? data.join("\n") : data;
//             const tokens = tokenize(d);
//             expect(tokens).toMatchSnapshot();
//         };
//         if (expectThrow) return expect(wrap).toThrowErrorMatchingSnapshot();
//         wrap();
//     });
// });

// const basicParserTests = [
//     ["parser test", "D=M;JMP"],
//     ["parser test 2", "@12"]
// ];

// basicParserTests.forEach(([tag, data, expectThrow]) => {
//     it(tag, () => {
//         const wrap = () => {
//             const d = data instanceof Array ? data.join("\n") : data;
//             const tokens = tokenize(d);
//             const ast = parser(tokens, d);
//             expect(ast).toMatchSnapshot()
//         };
//         if (expectThrow) return expect(wrap).toThrowErrorMatchingSnapshot();
//         wrap();
//     });
// });

it('test', ()=>{})