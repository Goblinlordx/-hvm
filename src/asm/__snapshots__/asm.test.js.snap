// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fail parsing number with decimal 1`] = `
"Invalid token, line: 1 col: 2
1.23
 ^
"
`;

exports[`parser test 1`] = `true`;

exports[`parser test 2`] = `
Object {
  "children": Array [],
  "type": "root",
}
`;

exports[`parser test 2 1`] = `
Object {
  "children": Array [],
  "type": "root",
}
`;

exports[`tokenize basic program 1`] = `
Array [
  Object {
    "col": 17,
    "idx": 16,
    "line": 1,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 12,
    "idx": 28,
    "line": 2,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 31,
    "line": 3,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 32,
    "line": 3,
    "type": "text",
    "value": "R.-;1+=",
  },
  Object {
    "col": 11,
    "idx": 39,
    "line": 3,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 42,
    "line": 4,
    "type": "text",
    "value": "D=A",
  },
  Object {
    "col": 6,
    "idx": 45,
    "line": 4,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 48,
    "line": 5,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 49,
    "line": 5,
    "type": "number",
    "value": 0,
  },
  Object {
    "col": 5,
    "idx": 50,
    "line": 5,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 53,
    "line": 6,
    "type": "text",
    "value": "M=D",
  },
  Object {
    "col": 6,
    "idx": 56,
    "line": 6,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 59,
    "line": 7,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 60,
    "line": 7,
    "type": "number",
    "value": 2,
  },
  Object {
    "col": 5,
    "idx": 61,
    "line": 7,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 64,
    "line": 8,
    "type": "text",
    "value": "D=A",
  },
  Object {
    "col": 6,
    "idx": 67,
    "line": 8,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 70,
    "line": 9,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 71,
    "line": 9,
    "type": "number",
    "value": 1,
  },
  Object {
    "col": 5,
    "idx": 72,
    "line": 9,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 75,
    "line": 10,
    "type": "text",
    "value": "M=D",
  },
  Object {
    "col": 6,
    "idx": 78,
    "line": 10,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 22,
    "idx": 100,
    "line": 11,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 1,
    "idx": 101,
    "line": 12,
    "type": "paren_open",
    "value": "(",
  },
  Object {
    "col": 2,
    "idx": 102,
    "line": 12,
    "type": "text",
    "value": "MAX",
  },
  Object {
    "col": 5,
    "idx": 105,
    "line": 12,
    "type": "paren_close",
    "value": ")",
  },
  Object {
    "col": 6,
    "idx": 106,
    "line": 12,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 109,
    "line": 13,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 110,
    "line": 13,
    "type": "text",
    "value": "R_0",
  },
  Object {
    "col": 7,
    "idx": 113,
    "line": 13,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 116,
    "line": 14,
    "type": "text",
    "value": "D=M",
  },
  Object {
    "col": 6,
    "idx": 119,
    "line": 14,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 122,
    "line": 15,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 123,
    "line": 15,
    "type": "text",
    "value": "R.1",
  },
  Object {
    "col": 7,
    "idx": 126,
    "line": 15,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 129,
    "line": 16,
    "type": "text",
    "value": "D=D-M",
  },
  Object {
    "col": 8,
    "idx": 134,
    "line": 16,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 137,
    "line": 17,
    "type": "text",
    "value": "D;JGT",
  },
  Object {
    "col": 8,
    "idx": 142,
    "line": 17,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 1,
    "idx": 143,
    "line": 18,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 146,
    "line": 19,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 147,
    "line": 19,
    "type": "neg",
    "value": "-",
  },
  Object {
    "col": 5,
    "idx": 148,
    "line": 19,
    "type": "number",
    "value": 123,
  },
  Object {
    "col": 8,
    "idx": 151,
    "line": 19,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 154,
    "line": 20,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 155,
    "line": 20,
    "type": "neg",
    "value": "-",
  },
  Object {
    "col": 5,
    "idx": 156,
    "line": 20,
    "type": "text",
    "value": "test",
  },
  Object {
    "col": 9,
    "idx": 160,
    "line": 20,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 3,
    "idx": 163,
    "line": 21,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 4,
    "idx": 164,
    "line": 21,
    "type": "number",
    "value": 0,
  },
  Object {
    "col": 5,
    "idx": 165,
    "line": 21,
    "type": "new_line",
    "value": "
",
  },
]
`;

exports[`tokenize instruction as text 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "text",
    "value": "D=A-1",
  },
]
`;

exports[`tokenize skips basic comment with whitespace 1`] = `Array []`;

exports[`tokenize skips comment 1`] = `Array []`;

exports[`tokenizes @ symbol 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "at_symbol",
    "value": "@",
  },
]
`;

exports[`tokenizes @ symbol with name 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "at_symbol",
    "value": "@",
  },
  Object {
    "col": 2,
    "idx": 1,
    "line": 1,
    "type": "text",
    "value": "S.FS-C;",
  },
]
`;

exports[`tokenizes basic comment in with empty lines 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 16,
    "idx": 16,
    "line": 2,
    "type": "new_line",
    "value": "
",
  },
  Object {
    "col": 15,
    "idx": 31,
    "line": 3,
    "type": "new_line",
    "value": "
",
  },
]
`;

exports[`tokenizes neg 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "neg",
    "value": "-",
  },
]
`;

exports[`tokenizes new lines 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "new_line",
    "value": "
",
  },
]
`;

exports[`tokenizes positive numeric value 1`] = `
Array [
  Object {
    "col": 1,
    "idx": 0,
    "line": 1,
    "type": "number",
    "value": 123,
  },
]
`;

exports[`tokenizes positive numeric value with whitespace 1`] = `
Array [
  Object {
    "col": 2,
    "idx": 1,
    "line": 1,
    "type": "number",
    "value": 123,
  },
]
`;
